File: cpp_merge/packages/integration/otc_agent/src/utils/price-validator.cpp
#include "price-validator.h"

std::shared_ptr<Promise<std::shared_ptr<PriceValidationResult>>> checkPriceDivergence(string tokenAddress, std::shared_ptr<Chain> chain, double poolPriceUsd)
{
    if (OR((!poolPriceUsd), (poolPriceUsd <= 0))) {
        return object{
            object::pair{std:("valid"), true}
        };
    }
    auto platformId = const_(COINGECKO_CHAIN_MAP)[chain];
    if (!platformId) {
        return object{
            object::pair{std:("valid"), true}
        };
    }
    try
    {
        auto url = std:("https://api.coingecko.com/api/v3/simple/token_price/") + platformId + std:("?contract_addresses=") + tokenAddress + std:("&vs_currencies=usd");
        auto cacheKey = std:("coingecko:") + platformId + std:(":") + tokenAddress->toLowerCase() + string_empty;
        auto data = std::async([=]() { fetchJsonWithRetryAndCache<std::shared_ptr<CoinGeckoPriceResponse>>(url, object{
            object::pair{std:("headers"), object{
                object::pair{std:("Accept"), std:("application/json")}
            }}
        }, object{
            object::pair{std:("cacheTtlMs"), COINGECKO_CACHE_TTL_MS}, 
            object::pair{std:("cacheKey"), std:("cacheKey")}, 
            object::pair{std:("maxRetries"), 3}
        }); });
        auto tokenData = (*const_(data))[tokenAddress->toLowerCase()];
        if (OR((!tokenData), (!tokenData["usd"]))) {
            return object{
                object::pair{std:("valid"), true}
            };
        }
        auto aggregatedPrice = tokenData["usd"];
        auto diff = Math->abs(poolPriceUsd - aggregatedPrice);
        auto divergence = diff / aggregatedPrice;
        auto divergencePercent = divergence * 100;
        if (divergencePercent > 10) {
            return object{
                object::pair{std:("valid"), false}, 
                object::pair{std:("warning"), std:("Price Warning: Pool price ($") + poolPriceUsd->toFixed(4) + std:(") diverges by ") + divergencePercent->toFixed(1) + std:("% from market price ($") + aggregatedPrice->toFixed(4) + std:(").")}, 
                object::pair{std:("aggregatedPrice"), std:("aggregatedPrice")}, 
                object::pair{std:("poolPrice"), poolPriceUsd}, 
                object::pair{std:("divergencePercent"), std:("divergencePercent")}
            };
        }
        return object{
            object::pair{std:("valid"), true}, 
            object::pair{std:("aggregatedPrice"), std:("aggregatedPrice")}, 
            object::pair{std:("poolPrice"), poolPriceUsd}, 
            object::pair{std:("divergencePercent"), std:("divergencePercent")}
        };
    }
    catch (const any& error)
    {
        console->warn(std:("Price validation failed:"), error);
        return object{
            object::pair{std:("valid"), true}
        };
    }
};


Record<string, string> COINGECKO_CHAIN_MAP = object{
    object::pair{std:("base"), std:("base")}, 
    object::pair{std:("solana"), std:("solana")}, 
    object::pair{std:("bsc"), std:("binance-smart-chain")}, 
    object::pair{std:("ethereum"), std:("ethereum")}
};
double COINGECKO_CACHE_TTL_MS = 30000;

void Main(void)
{
}

MAIN

File: otc-agent/src/utils/price-validator.ts
import { Chain } from "@/config/chains";
import { fetchJsonWithRetryAndCache } from "./retry-cache";

interface PriceValidationResult {
  valid: boolean;
  warning?: string;
  aggregatedPrice?: number;
  poolPrice?: number;
  divergencePercent?: number;
}

interface CoinGeckoPriceResponse {
  [address: string]: {
    usd?: number;
  };
}

// CoinGecko platform IDs for supported chains
// See: https://www.coingecko.com/en/api/documentation (asset_platforms endpoint)
const COINGECKO_CHAIN_MAP: Record<string, string> = {
  base: "base",
  solana: "solana",
  bsc: "binance-smart-chain",
  ethereum: "ethereum",
};

// Cache TTL for CoinGecko prices (30 seconds)
const COINGECKO_CACHE_TTL_MS = 30_000;

/**
 * Check if the pool price diverges significantly (>10%) from the aggregated off-chain price.
 * Uses CoinGecko public API with exponential retry and 30-second caching.
 */
export async function checkPriceDivergence(
  tokenAddress: string,
  chain: Chain,
  poolPriceUsd: number,
): Promise<PriceValidationResult> {
  if (!poolPriceUsd || poolPriceUsd <= 0) {
    return { valid: true }; // Cannot validate if pool price is invalid
  }

  const platformId = COINGECKO_CHAIN_MAP[chain];
  if (!platformId) {
    return { valid: true }; // Unsupported chain for validation
  }

  try {
    // Fetch price from CoinGecko with retry and caching
    const url = `https://api.coingecko.com/api/v3/simple/token_price/${platformId}?contract_addresses=${tokenAddress}&vs_currencies=usd`;
    const cacheKey = `coingecko:${platformId}:${tokenAddress.toLowerCase()}`;

    const data = await fetchJsonWithRetryAndCache<CoinGeckoPriceResponse>(
      url,
      { headers: { Accept: "application/json" } },
      {
        cacheTtlMs: COINGECKO_CACHE_TTL_MS,
        cacheKey,
        maxRetries: 3,
      },
    );

    const tokenData = data[tokenAddress.toLowerCase()];

    if (!tokenData || !tokenData.usd) {
      return { valid: true }; // No price found off-chain
    }

    const aggregatedPrice = tokenData.usd;

    // Calculate divergence
    const diff = Math.abs(poolPriceUsd - aggregatedPrice);
    const divergence = diff / aggregatedPrice;
    const divergencePercent = divergence * 100;

    if (divergencePercent > 10) {
      return {
        valid: false,
        warning: `Price Warning: Pool price ($${poolPriceUsd.toFixed(4)}) diverges by ${divergencePercent.toFixed(1)}% from market price ($${aggregatedPrice.toFixed(4)}).`,
        aggregatedPrice,
        poolPrice: poolPriceUsd,
        divergencePercent,
      };
    }

    return {
      valid: true,
      aggregatedPrice,
      poolPrice: poolPriceUsd,
      divergencePercent,
    };
  } catch (error) {
    console.warn("Price validation failed:", error);
    return { valid: true }; // Fail open
  }
}
